zpages/rpcz.go

```diff
@@ -205,113 +189,112 @@ func (s snapExporter) ExportView(vd *view.Data) {
 	if len(vd.Rows) == 0 {
 		return
 	}
-	ageSec := float64(time.Now().Sub(programStartTime)) / float64(time.Second)
 
-	computeRate := func(maxSec, x float64) float64 {
-		dur := ageSec
-		if maxSec > 0 && dur > maxSec {
-			dur = maxSec
-		}
-		return x / dur
-	}
+	mu.Lock()
+	defer mu.Unlock()
 
-	convertTime := func(ms float64) time.Duration {
-		if math.IsInf(ms, 0) || math.IsNaN(ms) {
-			return 0
+	if vd.View == ocgrpc.ServerCompletedRPCsView || vd.View == ocgrpc.ClientCompletedRPCsView {
+		errors := make(map[string]int64)
+		for _, row := range vd.Rows {
+			method := getMethod(row)
+			if method == "" {
+				continue
+			}
+			status := getStatus(row)
+			if status != "" && status != "OK" {
+				errors[method]++
+			}
 		}
-		return time.Duration(float64(time.Millisecond) * ms)
+		for method, errorCount := range errors {
+			s := snapshot(method, received)
+			dist := &view.DistributionData{
+				Mean:  1.0,
+				Count: errorCount,
+			}
+			s.ErrorsHour.update(vd.End, dist)
+			s.ErrorsMinute.update(vd.End, dist)
+		}
+		return
 	}
 
-	haveResetErrors := make(map[string]struct{})
-
-	mu.Lock()
-	defer mu.Unlock()
 	for _, row := range vd.Rows {
-		var method string
-		for _, tag := range row.Tags {
-			if tag.Key == ocgrpc.KeyClientMethod || tag.Key == ocgrpc.KeyServerMethod {
-				method = tag.Value
-				break
-			}
+		method := getMethod(row)
+		if method == "" {
+			continue
 		}
 
-		key := methodKey{method: method, received: received}
-		s := snaps[key]
-		if s == nil {
-			s = &statSnapshot{Method: method, Received: received}
-			snaps[key] = s
-		}
+		s := snapshot(method, received)
 
 		var (
-			sum   float64
-			count float64
+			dist *view.DistributionData
+			ok   bool
 		)
-		switch v := row.Data.(type) {
-		case *view.CountData:
-			sum = float64(v.Value)
-			count = float64(v.Value)
-		case *view.DistributionData:
-			sum = v.Sum()
-			count = float64(v.Count)
-		case *view.SumData:
-			sum = v.Value
-			count = v.Value
+		if dist, ok = row.Data.(*view.DistributionData); !ok {
+			continue
 		}
 
-		// Update field of s corresponding to the view.
+		var hour, minute *windowStat
 		switch vd.View {
-		case ocgrpc.ClientCompletedRPCsView:
-			if _, ok := haveResetErrors[method]; ok {
-				haveResetErrors[method] = struct{}{}
-				s.ErrorsTotal = 0
-			}
-			for _, tag := range row.Tags {
-				if tag.Key == ocgrpc.KeyClientStatus && tag.Value != "OK" {
-					s.ErrorsTotal += int(count)
-				}
-			}
-
 		case ocgrpc.ClientRoundtripLatencyView:
-			s.AvgLatencyTotal = convertTime(sum / count)
-
+			hour = &s.LatencyHour
+			minute = &s.LatencyMinute
 		case ocgrpc.ClientSentBytesPerRPCView:
```

Comment:
It might aid with readability if you could couple the outputs and rates together instead of a new case clause for each view i.e.
```go
case ocgrpc.ClientSentBytesPerRPCView, ocgrpc.ServerCompletedRPCsView:
       hour = &s.OutputHour
       minute = &s.OutputMinute
case ocgrpc.ClientLatencyView, ocgrpc.ServerLatencyView:
       hour = &s.LatencyHour
       minute = &s.LatencyMinute
```

Reply:
yes! much better thanks